### PR TITLE
Add a whole corpus sampler

### PR DESCRIPTION
**Pull request by boomanaiden154**
This patch adds a whole corpus sampler. This is intended to be used for trace data collection where we need to compile an entire corpus subset that is extracted using separate tooling in order to generate a reward.

### CODE REVIEW

File: compiler_opt/rl/corpus.py [hunk 1]
```diff
@@ -224,6 +224,40 @@ def __call__(self, k: int, n: int = 10) -> List[ModuleSpec]:
     return list(results)
 
 
+class WholeCorpusSampler(Sampler):
+  """Returns the entire corpus every time a sample is requested."""
```
Comment (mtrofin) [inline compiler_opt/rl/corpus.py:228]: is the the entire corpus, or it's that subset of IR modules that cover a function graph?

File: compiler_opt/rl/corpus.py [hunk 2]
```diff
@@ -224,6 +224,40 @@ def __call__(self, k: int, n: int = 10) -> List[ModuleSpec]:
...
+  # as super().__init__ is an abstract method that we need to override to
+  # execute it without throwing an error.
+  def __init__(self, module_specs: Tuple[ModuleSpec]):  # pylint: disable=useless-super-delegation
+    super().__init__(module_specs)
+
+  def reset(self):
+    pass
+
+  def __call__(self, k: int, n: int = 10) -> List[ModuleSpec]:
+    """Returns the entire corpus a list of module specs.
```
Comment (mtrofin) [inline compiler_opt/rl/corpus.py:240]: nit: is there a comma missing?

File: compiler_opt/rl/corpus.py [hunk 3]
```diff
@@ -224,6 +224,40 @@ def __call__(self, k: int, n: int = 10) -> List[ModuleSpec]:
...
+
+    Args:
+      k: The number of modules to sample. This must be equal to the number
```
Comment (mtrofin) [inline compiler_opt/rl/corpus.py:243]: why do we even need "k" here?

File: compiler_opt/rl/corpus.py (hunk 1 above)
Comment (boomanaiden154) [reply compiler_opt/rl/corpus.py:228]: It is the subset of IR modules that cover a trace. That is in effect a whole corpus though given the tooling creates a new `corpus_description.json` for the corpus subset.

File: compiler_opt/rl/corpus.py (hunk 2 above)
Comment (boomanaiden154) [reply compiler_opt/rl/corpus.py:240]: Missing a word.

File: compiler_opt/rl/corpus.py (hunk 3 above)
Comment (boomanaiden154) [reply compiler_opt/rl/corpus.py:243]: We have to include `k` and `n` or otherwise we get lint errors given the base class has those as parameters.

Comment (mtrofin) [reply compiler_opt/rl/corpus.py:243]: another way to understand the problem, though, would be to say that given an initial module-level corpus, and a set of subsets of modules, you get a new corpus where a corpus element is such a subset of modules.

then "k" counts how many of those subsets.

Comment (boomanaiden154) [reply compiler_opt/rl/corpus.py:243]: Sure, but I don't plan on supporting multiple entrypoints to begin with. That would also require more stuff passed around here as we would additionally need to pass the trace associated with that entrypoint associated with this corpus subset.

**Comment by boomanaiden154** [comment]
This is unnecessary when using a custom `BlackboxEvaluator`.

**Pull request closed without merging**